packages/@cdktf/hcl2json/main.go

```diff
@@ -159,3 +191,193 @@ func literalValueExpr(val cty.Value) (hclsyntax.Expression, hcl.Diagnostics) {
+	hclsyntax.OpLogicalNot:         "LogicalNot",
+	hclsyntax.OpEqual:              "Equal",
+	hclsyntax.OpNotEqual:           "NotEqual",
+	hclsyntax.OpGreaterThan:        "GreaterThan",
+	hclsyntax.OpGreaterThanOrEqual: "GreaterThanOrEqual",
+	hclsyntax.OpLessThan:           "LessThan",
+	hclsyntax.OpLessThanOrEqual:    "LessThanOrEqual",
+	hclsyntax.OpAdd:                "Add",
+	hclsyntax.OpSubtract:           "Subtract",
+	hclsyntax.OpMultiply:           "Multiply",
+	hclsyntax.OpDivide:             "Divide",
+	hclsyntax.OpModulo:             "Modulo",
+	hclsyntax.OpNegate:             "Negate",
+}
+
+type ExpressionAst struct {
+	Parent   *ExpressionAst         `json:"-"`
+	Children []*ExpressionAst       `json:"children"`
+	Type     string                 `json:"type"`
+	Meta     map[string]interface{} `json:"meta,omitempty"`
+	Range    RangeAst               `json:"range,omitempty"`
+}
+
+type PosAst struct {
+	Line   int `json:"line"`
+	Column int `json:"column"`
+	Byte   int `json:"byte"`
+}
+type RangeAst struct {
+	Start PosAst `json:"start"`
+	End   PosAst `json:"end"`
+}
+
+type TraversalAst struct {
+	Segment string    `json:"segment"`
+	Range   hcl.Range `json:"range,omitempty"`
+}
+
+type ExpressionWalker struct {
+	Root    ExpressionAst
+	Stack   []*ExpressionAst
+	Current *ExpressionAst
+}
+
+func convertPosToAst(p hcl.Pos) PosAst {
+	return PosAst{
+		Line:   p.Line,
+		Column: p.Column,
+		Byte:   p.Byte,
+	}
+}
+
+func convertRangeToAst(r hcl.Range) RangeAst {
+	return RangeAst{
+		Start: convertPosToAst(r.Start),
+		End:   convertPosToAst(r.End),
+	}
+}
+
+func convertTraversal(t hcl.Traversal) []TraversalAst {
+	var res []TraversalAst
+	for _, p := range t {
+		switch p.(type) {
+		case hcl.TraverseRoot:
+			root := p.(hcl.TraverseRoot)
+			res = append(res, TraversalAst{Segment: root.Name, Range: root.SrcRange})
+		case hcl.TraverseAttr:
+			attr := p.(hcl.TraverseAttr)
+			res = append(res, TraversalAst{Segment: attr.Name, Range: attr.SrcRange})
+		}
+	}
+	return res
+}
+
+func getOperationName(op *hclsyntax.Operation) string {
+	return operationNamesLookup[op]
+}
+
+func (w *ExpressionWalker) Enter(node hclsyntax.Node) hcl.Diagnostics {
+	expr := node.(hclsyntax.Expression)
+	if w.Current != nil {
+		w.Stack = append(w.Stack, w.Current)
+		newAst := ExpressionAst{
+			Parent:   w.Current,
+			Children: []*ExpressionAst{},
+			Meta:     map[string]interface{}{},
+		}
+		w.Current.Children = append(w.Current.Children, &newAst)
+		w.Current = &newAst
+	} else {
+		w.Root = ExpressionAst{
+			Parent:   nil,
+			Children: []*ExpressionAst{},
+			Meta:     map[string]interface{}{},
+		}
+		w.Current = &w.Root
+	}
+
+	w.Current.Range = convertRangeToAst(expr.StartRange())
+
+	switch expr.(type) {
+	case *hclsyntax.FunctionCallExpr:
+		fnExpr := expr.(*hclsyntax.FunctionCallExpr)
+		w.Current.Type = "function"
+		w.Current.Meta["name"] = fnExpr.Name
+		w.Current.Meta["expandedFinalArgument"] = fnExpr.ExpandFinal
```

Comment:
What is this? I'm mainly curious as the name is changed from "expand" to "expanded" – is this lazily loaded when accessing `.ExpandFinal`?

It might be nice to have TS types for meta which could feature an explanation what each of these are – but that can go in a follow-up PR as well.

Reply:
Yeah, once I'm done reviewing all the properties, I'll do a TS type for the AST